File: core/simplify.py
# -*- coding: utf-8 -*-
# pylint: disable=no-name-in-module
# pylint: disable=too-many-lines
# pylint: disable=useless-return
# pylint: disable=too-few-public-methods
# pylint: disable=relative-beyond-top-level

from qgis.core import QgsLineString, QgsWkbTypes, QgsGeometry

from ..geo_sim_util import Epsilon, GsCollection, GeoSimUtil, GsFeature, ProgressBar


# --------------------------------------------------------
# Start of the algorithm
# --------------------------------------------------------

# Define global constant


class RbResults:
    """Class defining the stats and results"""

    __slots__ = ('in_nbr_features', 'out_nbr_features', 'nbr_vertice_deleted',  'qgs_features_out', 'nbr_pass',
                 'is_structure_valid')

    def __init__(self):
        """Constructor that initialize a RbResult object.

        :param: None
        :return: None
        :rtype: None
        """

        self.in_nbr_features = None
        self.out_nbr_features = None
        self.nbr_vertice_deleted = 0
        self.qgs_features_out = None
        self.nbr_pass = 0
        self.is_structure_valid = None


class Simplify:
    """Main class for bend reduction"""

    @staticmethod
    def douglas_peucker(qgs_in_features, tolerance, validate_structure=False, feedback=None):
        """Main static method used to launch the simplification of the Douglas-Peucker algorithm.

        :param: qgs_features: List of QgsFeatures to process.
        :param: tolerance: Simplification tolerance in ground unit.
        :param: validate_structure: Validate internal data structure after processing (for debugging only)
        :param: feedback: QgsFeedback handle for interaction with QGIS.
        :return: Statistics and results object.
        :rtype: RbResults
        """

        dp = Simplify(qgs_in_features, tolerance, validate_structure, feedback)
        results = dp.reduce()

        return results

    @staticmethod
    def find_farthest_point(qgs_points, first, last, ):
        """Returns a tuple with the farthest point's index and it's distance from a subline section

        :param: qgs_points: List of QgsPoint defining the line to process
        :first: int: Index of the first point in qgs_points
        :last: int: Index of the last point in qgs_points
        :return: distance from the farthest point; index of the farthest point
        :rtype: tuple of 2 values
        """

        if last - first >= 2:
            qgs_geom_first_last = QgsLineString(qgs_points[first], qgs_points[last])
            qgs_geom_engine = QgsGeometry.createGeometryEngine(qgs_geom_first_last)
            distances = [qgs_geom_engine.distance(qgs_points[i]) for i in range(first + 1, last)]
            farthest_dist = max(distances)
            farthest_index = distances.index(farthest_dist) + first + 1
        else:
            # Not enough vertice to calculate the farthest distance
            farthest_dist = -1.
            farthest_index = first

        return farthest_index, farthest_dist

    __slots__ = ('tolerance', 'validate_structure', 'feedback', 'rb_collection', 'eps', 'rb_results', 'rb_geoms',
                 'gs_features')

    def __init__(self, qgs_in_features, tolerance, validate_structure, feedback):
        """Constructor for Simplify algorithm.

       :param: qgs_in_features: List of features to process.
       :param: tolerance: Float tolerance distance of the Douglas Peucker algorithm.
       :param: validate_structure: flag to validate internal data structure after processing (for debugging)
       :param: feedback: QgsFeedback handle for interaction with QGIS.
       """

        self.tolerance = tolerance
        self.validate_structure = validate_structure
        self.feedback = feedback

        # Calculates the epsilon and initialize some stats and results value
        self.eps = Epsilon(qgs_in_features)
        self.eps.set_class_variables()
        self.rb_results = RbResults()

        # Create the list of GsPolygon, GsLineString and GsPoint to process
        self.rb_results.in_nbr_features = len(qgs_in_features)
        self.gs_features = GsFeature.create_gs_feature(qgs_in_features)

    def reduce(self):
        """Main method to reduce line string.

        :return: Statistics and result object.
        :rtype: RbResult
        """

        #  Code used for the profiler (uncomment if needed)
#        import cProfile, pstats, io
#        from pstats import SortKey
#        pr = cProfile.Profile()
#        pr.enable()

        # Calculates the epsilon and initialize some stats and results value
#        self.eps = Epsilon(self.qgs_in_features)
#        self.eps.set_class_variables()
#        self.rb_results = RbResults()

#        # Create the list of GsPolygon, GsLineString and GsPoint to process
#        self.gs_features = GeoSimUtil.create_gs_feature(self.qgs_in_features)

        # Pre process the LineString: remove to close point and co-linear points
        self.rb_geoms = self.pre_simplification_process()

        # Create the GsCollection a spatial index to accelerate search for spatial relationships
        self.rb_collection = GsCollection()
        self.rb_collection.add_features(self.rb_geoms, self.feedback)

        # Execute the line simplification for each LineString
        self._simplify_lines()

        # Recreate the QgsFeature
        qgs_features_out = [gs_feature.get_qgs_feature() for gs_feature in self.gs_features]

        # Set return values
        self.rb_results.out_nbr_features = len(qgs_features_out)
        self.rb_results.qgs_features_out = qgs_features_out

        # Validate inner spatial structure. For debug purpose only
        if self.rb_results.is_structure_valid:
            self.rb_collection.validate_integrity(self.rb_geoms)

        #  Code used for the profiler (uncomment if needed)
 #       pr.disable()
 #       s = io.StringIO()
 #       sortby = SortKey.CUMULATIVE
 #       ps = pstats.Stats(pr, stream=s).sort_stats(sortby)
 #       ps.print_stats()
 #       print(s.getvalue())

        return self.rb_results

    def pre_simplification_process(self):
        """This method execute the pre simplification process

        Pre simplification process applies only to closed line string and is used to find the 2 points that are
        the distant from each other using the oriented bounding box

        :return: List of rb_geom
        :rtype: [RbGeom]
        """

        # Create the list of RbGeom ==> List of geometry to simplify
        sim_geoms = []
        for gs_feature in self.gs_features:
            sim_geoms += gs_feature.get_rb_geom()

        return sim_geoms

    def _simplify_lines(self):
        """Loop over the geometry until there is no more subline to simplify

        An iterative process for line simplification is applied in order to maximise line simplification.  The process
        will always stabilize and exit when there are no more simplification to do.
        """

        while True:
            progress_bar_value = 0
            self.rb_results.nbr_pass += 1
            progress_bar = ProgressBar(self.feedback, len(self.rb_geoms),
                                       "Iteration: {0}".format(self.rb_results.nbr_pass))
            nbr_vertice_deleted = 0
            for i, rb_geom in enumerate(self.rb_geoms):
                if self.feedback.isCanceled():
                    break
                progress_bar.set_value(i)
                if not rb_geom.is_simplest:  # Only process geometry that are not at simplest form
                    nbr_vertice_deleted += self.process_line(rb_geom)

            self.feedback.pushInfo("Vertice deleted: {0}".format(nbr_vertice_deleted))

            # While loop breaking condition (when no vertice deleted in a loop)
            if nbr_vertice_deleted == 0:
                break
            self.rb_results.nbr_vertice_deleted += nbr_vertice_deleted

        return

    def validate_constraints(self, sim_geom, first, last):
        """Validate the spatial relationship in order maintain topological structure

        Three distinct spatial relation are tested in order to assure that each bend reduce will continue to maintain
        the topological structure in a feature between the features:
         - Simplicity: Adequate validation is done to make sure that the bend reduction will not cause the feature
                       to cross  itself.
         - Intersection : Adequate validation is done to make sure that a line from other features will not intersect
                          the bend being reduced
         - Sidedness: Adequate validation is done to make sure that a line is not completely contained in the bend.
                      This situation can happen when a ring in a polygon complete;y lie in a bend ans after bend
                      reduction, the the ring falls outside the polygon which make it invalid.

        Note if the topological structure is wrong before the bend correction no correction will be done on these
        errors.

        :param: sim_geom: Geometry used to validate constraints
        :param: first: Index of the start vertice of the subline
        :param: last: Index of the last vertice of the subline
        :return: Flag indicating if the spatial constraints are valid for this subline simplification
        :rtype: Bool
        """

        constraints_valid = True

        qgs_points = [sim_geom.qgs_geom.vertexAt(i) for i in range(first, last+1)]
        qgs_geom_new_subline = QgsGeometry(QgsLineString(qgs_points[0], qgs_points[-1]))
        qgs_geom_old_subline = QgsGeometry(QgsLineString(qgs_points))
        qgs_geoms_with_itself, qgs_geoms_with_others = \
            self.rb_collection.get_segment_intersect(sim_geom.id, qgs_geom_old_subline.boundingBox(),
                                                     qgs_geom_old_subline)

        # First: check if the bend reduce line string is an OGC simple line
        constraints_valid = GeoSimUtil.validate_simplicity(qgs_geoms_with_itself, qgs_geom_new_subline)

        # Second: check that the new line does not intersect with any other line or points
        if constraints_valid and len(qgs_geoms_with_others) >= 1:
            constraints_valid = GeoSimUtil.validate_intersection(qgs_geoms_with_others, qgs_geom_new_subline)

        # Third: check that inside the subline to simplify there is no feature completely inside it.  This would cause a
        # sidedness or relative position error
        if constraints_valid and len(qgs_geoms_with_others) >= 1:
            qgs_ls_old_subline = QgsLineString(qgs_points)
            qgs_ls_old_subline.addVertex(qgs_points[0])  # Close the line with the start point
            qgs_geom_old_subline = QgsGeometry(qgs_ls_old_subline.clone())

            # Next two lines used to transform a self intersecting line into a valid MultiPolygon
            qgs_geom_unary = QgsGeometry.unaryUnion([qgs_geom_old_subline])
            qgs_geom_polygonize = QgsGeometry.polygonize([qgs_geom_unary])

            if qgs_geom_polygonize.isSimple():
                constraints_valid = GeoSimUtil.validate_sidedness(qgs_geoms_with_others, qgs_geom_polygonize)
            else:
                print("Polygonize not valid")
                constraints_valid = False

        return constraints_valid

    @staticmethod
    def init_process_line_stack(is_line_closed, qgs_points):
        """Method that initialize the stack used to simulate recursivity to simplify the line

        :param: is_closed: Boolean to indicate if the feature is closed or open
        :param: qgs_points: List of QgsPoints forming the line string to simplify
        :return: Stack used to initiate the line simplification process
        :rtype: List of tuple
        """

        stack = []
        last_index = len(qgs_points) - 1
        if is_line_closed:
            # Initialize stack for a closed line string
            if last_index >= 4:
                x = qgs_points[0].x()
                y = qgs_points[0].y()
                lst_distance = [qgs_point.distance(x, y) for qgs_point in qgs_points]
                mid_index = lst_distance.index(max(lst_distance))  # Most distant vertex position

                (farthest_index_a, farthest_dist_a) = Simplify.find_farthest_point(qgs_points, 0, mid_index)
                (farthest_index_b, farthest_dist_b) = Simplify.find_farthest_point(qgs_points, mid_index, last_index)
                if farthest_dist_a > 0.:
                    stack.append((0, farthest_index_a))
                    stack.append((farthest_index_a, mid_index))
                if farthest_dist_b > 0.:
                    stack.append((mid_index, farthest_index_b))
                    stack.append((farthest_index_b, last_index))
            else:
                # Not enough vertice... nothing to simplify
                pass
        else:
            # Initialize stack for an open line string
            stack.append((0, last_index))

        return stack

    def process_line(self, sim_geom):
        """This method is simplifying a line with the Douglas Peucker algorithm and spatial constraints.

        Important note: The line is always simplified for the end of the line to the start of the line. This helps
        maintain the relative position of the vertice in the line

        :param: sim_geom: GeoSim object to simplify
        :return: Number of vertice deleted
        :rtype: int
        """

        qgs_line_string = sim_geom.qgs_geom.constGet()
        qgs_points = qgs_line_string.points()

        # Initialize the stack that simulate recursivity
        stack = Simplify.init_process_line_stack(qgs_line_string.isClosed(), qgs_points)

        # Loop over the stack to simplify the line
        sim_geom.is_simplest = True
        nbr_vertice_deleted = 0
        while stack:
            (first, last) = stack.pop()
            if first + 1 < last:  # The segment to check has only 2 points
                (farthest_index, farthest_dist) = Simplify.find_farthest_point(qgs_points, first, last)
                if farthest_dist <= self.tolerance:
                    if self.validate_constraints(sim_geom, first, last):
                        nbr_vertice_deleted += last - first - 1
                        self.rb_collection.delete_vertex(sim_geom, first + 1, last - 1)
                    else:
                        sim_geom.is_simplest = False  # The line string is not at its simplest form
                        # In case of non respect of spatial constraints split and stack again the sub lines
                        (farthest_index, farthest_dist) = Simplify.find_farthest_point(qgs_points, first, last)
                        if farthest_dist <= self.tolerance:
                            # Stack for the net iteration
                            stack.append((first, farthest_index))
                            stack.append((farthest_index, last))
                else:
                    # Stack for the iteration
                    stack.append((first, farthest_index))
                    stack.append((farthest_index, last))

        return nbr_vertice_deleted

File: rust/src/simplify.rs
use anyhow::{anyhow, Result};

use crate::constraints::{validate_intersection, validate_sidedness, validate_simplicity};
use crate::epsilon::Epsilon;
use crate::geometry::{Coord, FeatureRecord, GsFeature, RbGeom, SimpleGeometry};
use crate::spatial_index::{bbox_for_coords, GsCollection};

#[derive(Debug, Clone, Copy)]
pub struct SimplifyParams {
    pub tolerance: f64,
    pub validate_structure: bool,
}

#[derive(Debug, Clone)]
pub struct SimplifyStats {
    pub in_nbr_features: usize,
    pub out_nbr_features: usize,
    pub nbr_vertice_deleted: usize,
    pub nbr_pass: usize,
    pub is_structure_valid: Option<bool>,
}

#[derive(Debug, Clone)]
pub struct SimplifyOutput {
    pub features: Vec<FeatureRecord>,
    pub stats: SimplifyStats,
}

#[derive(Debug, Clone, Copy)]
pub struct SimplifyProgress {
    pub current_feature: usize,
    pub total_features: usize,
    pub current_pass: usize,
}

pub struct SimplifyEngine {
    tolerance: f64,
    validate_structure: bool,
    eps: Epsilon,
    collection: GsCollection,
    gs_features: Vec<GsFeature>,
    rb_geoms: Vec<RbGeom>,
}

impl SimplifyEngine {
    pub fn new(records: &[FeatureRecord], params: SimplifyParams) -> Result<Self> {
        let eps = Epsilon::from_features(records);
        let (gs_features, rb_geoms) = GsFeature::from_records(records, eps.zero_relative)?;

        let mut collection = GsCollection::new(eps.zero_relative);
        collection.add_features(&rb_geoms);

        Ok(Self {
            tolerance: params.tolerance,
            validate_structure: params.validate_structure,
            eps,
            collection,
            gs_features,
            rb_geoms,
        })
    }

    pub fn run(mut self) -> Result<SimplifyOutput> {
        self.run_with_progress(|_| {})
    }

    pub fn run_with_progress<F>(mut self, mut on_progress: F) -> Result<SimplifyOutput>
    where
        F: FnMut(SimplifyProgress),
    {
        let in_count = self.gs_features.len();
        let mut total_deleted = 0usize;
        let mut pass = 0usize;

        loop {
            pass += 1;
            let mut deleted_in_pass = 0usize;

            for i in 0..self.rb_geoms.len() {
                on_progress(SimplifyProgress {
                    current_feature: i + 1,
                    total_features: self.rb_geoms.len(),
                    current_pass: pass,
                });

                if !self.rb_geoms[i].is_simplest {
                    deleted_in_pass += self.process_line(i)?;
                }
            }

            if deleted_in_pass == 0 {
                break;
            }
            total_deleted += deleted_in_pass;
        }

        let mut out_features = self
            .gs_features
            .iter()
            .map(|f| f.rebuild_record(&self.rb_geoms))
            .collect::<Result<Vec<_>>>()?;

        out_features.sort_by_key(|f| (f.stable_id, f.fid.unwrap_or(u64::MAX)));

        let structure_valid = if self.validate_structure {
            Some(self.collection.validate_integrity(&self.rb_geoms)?)
        } else {
            None
        };

        let stats = SimplifyStats {
            in_nbr_features: in_count,
            out_nbr_features: out_features.len(),
            nbr_vertice_deleted: total_deleted,
            nbr_pass: pass,
            is_structure_valid: structure_valid,
        };

        Ok(SimplifyOutput {
            features: out_features,
            stats,
        })
    }

    fn process_line(&mut self, rb_geom_idx: usize) -> Result<usize> {
        let mut stack = init_process_line_stack(self.rb_geoms[rb_geom_idx].is_closed(), &self.rb_geoms[rb_geom_idx].coords);

        let mut deleted = 0usize;
        self.rb_geoms[rb_geom_idx].is_simplest = true;

        while let Some((first, last)) = stack.pop() {
            if first + 1 >= last {
                continue;
            }

            let coords_snapshot = self.rb_geoms[rb_geom_idx].coords.clone();
            if last >= coords_snapshot.len() {
                continue;
            }

            let (farthest_index, farthest_dist) = find_farthest_point(&coords_snapshot, first, last);

            if farthest_dist <= self.tolerance {
                if self.validate_constraints(rb_geom_idx, first, last)? {
                    deleted += last - first - 1;
                    let rb_geom = self
                        .rb_geoms
                        .get_mut(rb_geom_idx)
                        .ok_or_else(|| anyhow!("invalid rb geom index"))?;
                    self.collection.delete_vertex(rb_geom, first + 1, last - 1)?;
                } else {
                    self.rb_geoms[rb_geom_idx].is_simplest = false;
                    let now = &self.rb_geoms[rb_geom_idx].coords;
                    if first < now.len() && last < now.len() {
                        let (next_farthest, next_dist) = find_farthest_point(now, first, last);
                        if next_dist <= self.tolerance {
                            stack.push((first, next_farthest));
                            stack.push((next_farthest, last));
                        }
                    }
                }
            } else {
                stack.push((first, farthest_index));
                stack.push((farthest_index, last));
            }
        }

        Ok(deleted)
    }

    fn validate_constraints(&self, rb_geom_idx: usize, first: usize, last: usize) -> Result<bool> {
        let sim_geom = self
            .rb_geoms
            .get(rb_geom_idx)
            .ok_or_else(|| anyhow!("invalid rb geom index for constraint validation"))?;

        if last >= sim_geom.coords.len() {
            return Ok(false);
        }

        let sub_coords = sim_geom.coords[first..=last].to_vec();
        if sub_coords.len() < 2 {
            return Ok(false);
        }

        let new_subline = SimpleGeometry::LineString(vec![sub_coords[0], *sub_coords.last().unwrap()]).to_geos()?;
        let old_subline = SimpleGeometry::LineString(sub_coords.clone()).to_geos()?;
        let bbox = bbox_for_coords(&sub_coords);

        let (with_itself, with_others) = self
            .collection
            .get_segment_intersect(sim_geom.id, bbox, &old_subline)?;

        if !validate_simplicity(&with_itself, &new_subline, self.eps.zero_relative)? {
            return Ok(false);
        }

        if !with_others.is_empty() && !validate_intersection(&with_others, &new_subline)? {
            return Ok(false);
        }

        if !with_others.is_empty() {
            return validate_sidedness(&with_others, &sub_coords);
        }

        Ok(true)
    }
}

pub fn init_process_line_stack(is_line_closed: bool, points: &[Coord]) -> Vec<(usize, usize)> {
    let mut stack = Vec::new();
    if points.is_empty() {
        return stack;
    }

    let last_index = points.len() - 1;

    if is_line_closed {
        if last_index >= 4 {
            let origin = points[0];
            let mut max_d = f64::NEG_INFINITY;
            let mut mid_index = 0usize;
            for (idx, p) in points.iter().enumerate() {
                let dx = p.x - origin.x;
                let dy = p.y - origin.y;
                let d = (dx * dx + dy * dy).sqrt();
                if d > max_d {
                    max_d = d;
                    mid_index = idx;
                }
            }

            let (farthest_a, dist_a) = find_farthest_point(points, 0, mid_index);
            let (farthest_b, dist_b) = find_farthest_point(points, mid_index, last_index);

            if dist_a > 0.0 {
                stack.push((0, farthest_a));
                stack.push((farthest_a, mid_index));
            }
            if dist_b > 0.0 {
                stack.push((mid_index, farthest_b));
                stack.push((farthest_b, last_index));
            }
        }
    } else {
        stack.push((0, last_index));
    }

    stack
}

pub fn find_farthest_point(points: &[Coord], first: usize, last: usize) -> (usize, f64) {
    if last < first + 2 {
        return (first, -1.0);
    }

    let a = points[first];
    let b = points[last];
    let mut farthest_index = first;
    let mut farthest_dist = f64::NEG_INFINITY;

    for (idx, p) in points.iter().enumerate().take(last).skip(first + 1) {
        let d = point_to_segment_dist(*p, a, b);
        if d > farthest_dist {
            farthest_dist = d;
            farthest_index = idx;
        }
    }

    (farthest_index, farthest_dist)
}

pub fn point_to_segment_dist(p: Coord, a: Coord, b: Coord) -> f64 {
    let vx = b.x - a.x;
    let vy = b.y - a.y;
    let wx = p.x - a.x;
    let wy = p.y - a.y;

    let seg_len2 = vx * vx + vy * vy;
    if seg_len2 == 0.0 {
        let dx = p.x - a.x;
        let dy = p.y - a.y;
        return (dx * dx + dy * dy).sqrt();
    }

    let t = ((wx * vx) + (wy * vy)) / seg_len2;
    let t = t.clamp(0.0, 1.0);

    let proj_x = a.x + t * vx;
    let proj_y = a.y + t * vy;

    let dx = p.x - proj_x;
    let dy = p.y - proj_y;

    (dx * dx + dy * dy).sqrt()
}
